contracts/rewards/src/state.rs

```diff
@@ -0,0 +1,383 @@
+use std::collections::HashMap;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Addr, Storage, Uint256};
+use cw_storage_plus::{Item, Map};
+
+use crate::{error::ContractError, msg::RewardsParams};
+
+#[cw_serde]
+pub struct Config {
+    pub params: RewardsParams,
+}
+
+#[cw_serde]
+pub struct EpochTally {
+    pub epoch_num: u64,
+    pub contract: Addr,
+    pub total_events: u64,
+    pub participation: HashMap<Addr, u64>,
+    pub rewards_rate: Uint256,
+}
+
+impl EpochTally {
+    #[allow(dead_code)]
+    pub fn new(epoch_num: u64, contract: Addr, rewards_rate: Uint256) -> Self {
+        EpochTally {
+            epoch_num,
+            contract,
+            total_events: 0,
+            participation: HashMap::new(),
+            rewards_rate,
+        }
+    }
+}
+
+#[cw_serde]
+pub struct Event {
+    event_id: String,
+    contract: Addr,
+    epoch_num: u64,
+}
+
+#[cw_serde]
+pub struct Epoch {
+    pub epoch_num: u64,
+    pub block_height_started: u64,
+}
+
+#[cw_serde]
+pub struct RewardsPool {
+    contract: Addr,
+    balance: Uint256,
+}
+
+impl RewardsPool {
+    #[allow(dead_code)]
+    pub fn new(contract: Addr) -> Self {
+        RewardsPool {
+            contract,
+            balance: Uint256::zero(),
+        }
+    }
+}
+
+pub trait Store {
+    fn load_config(&self) -> Config;
+
+    fn load_last_checkpoint(&self) -> Result<Epoch, ContractError>;
+
+    fn load_rewards_watermark(&self) -> Result<Option<u64>, ContractError>;
+
+    fn load_event(&self, event_id: String, contract: Addr) -> Result<Option<Event>, ContractError>;
+
+    fn load_epoch_tally(
+        &self,
+        contract: Addr,
+        epoch_num: u64,
+    ) -> Result<Option<EpochTally>, ContractError>;
+
+    fn load_rewards_pool(&self, contract: Addr) -> Result<Option<RewardsPool>, ContractError>;
+
+    fn save_config(&mut self, config: &Config) -> Result<(), ContractError>;
+
+    fn save_last_checkpoint(&mut self, epoch: &Epoch) -> Result<(), ContractError>;
+
+    fn save_rewards_watermark(&mut self, epoch_num: u64) -> Result<(), ContractError>;
+
+    fn save_event(&mut self, event: &Event) -> Result<(), ContractError>;
+
+    fn save_epoch_tally(&mut self, tally: &EpochTally) -> Result<(), ContractError>;
+
+    fn save_rewards_pool(&mut self, pool: &RewardsPool) -> Result<(), ContractError>;
+}
+
+pub const CONFIG: Item<Config> = Item::new("config");
+
+const LAST_CHECKPOINT: Item<Epoch> = Item::new("current_epoch");
+
+const TALLIES: Map<(Addr, u64), EpochTally> = Map::new("tallies");
+
+const EVENTS: Map<(String, Addr), Event> = Map::new("events");
+
+const POOLS: Map<Addr, RewardsPool> = Map::new("pools");
+
+const WATERMARK: Item<u64> = Item::new("rewards_watermark");
```

Comment:
Can you add comments explaining what these are and what are the keys? What is the watermark?